upgrade/upgrade-scripts/upgrade

```diff
@@ -395,16 +397,21 @@ function finalize() {
 		;;
 	esac
 
-	ROOTFS_DATASET=$(get_mounted_rootfs_container_dataset)
-	[[ -n "$ROOTFS_DATASET" ]] ||
-		die "unable to determine mounted rootfs container dataset"
-
-	SNAPSHOT_NAME=$(get_dataset_rollback_snapshot_name "$ROOTFS_DATASET")
-	[[ -n "$SNAPSHOT_NAME" ]] ||
-		die "unable to determine rollback snapshot name"
-
-	zfs destroy -r "$ROOTFS_DATASET@$SNAPSHOT_NAME" ||
-		die "failed to destroy rollback snapshots"
+	SNAPSHOT_LIST=$(zfs list -t snapshot -r rpool/ROOT -d 2 -Ho name) ||
+		die "unable to determine available rollback snapshots"
+
+	for snapshot in $SNAPSHOT_LIST; do
+		grep -qE "^execute-upgrade.[[:alnum:]]{7}$" <(echo "$snapshot" | awk -F@ '{print $2}') ||
+			die "unexpected snapshot name: '$snapshot'"
+		local SNAPSHOT_VERSION
+		SNAPSHOT_VERSION=$(zfs get -Hpo value "$PROP_CURRENT_VERSION" "$snapshot") ||
```

Comment:
After setting `SNAPSHOT_VERSION`, we should probably verify a version was obtained, since by default it will successfully return `-`.. e.g.
```
[[ "$SNAPSHOT_VERSION" != "-" ]] || die "failed to get current version for snapshot '$snapshot'"
```

Reply:
I see, let me include that